test/controllers/observations_controller_test.rb

```diff
@@ -2850,36 +2852,34 @@ def test_image_upload_when_create_fails
   end
 
   def test_image_upload_when_process_image_fails
-    login("rolf")
-
     setup_image_dirs
     file = Rails.root.join("test/images/Coprinus_comatus.jpg")
     file = Rack::Test::UploadedFile.new(file, "image/jpeg")
+    image = Image.create(user: users(:rolf),
+                         copyright_holder: "zuul",
+                         when: Time.current,
+                         notes: "stubbed in test")
+    params = { observation: { place_name: "USA",
+                              when: Time.current },
+               image: { "0" => { image: file,
+                                 copyright_holder: "zuul",
+                                 when: Time.current } } }
+    login("rolf")
 
     # Simulate process_image failure.
-    Image.any_instance.stubs(:process_image).returns(false)
-
-    post(
-      :create,
-      params: {
-        observation: {
-          place_name: "USA",
-          when: Time.current
-        },
-        image: {
-          "0" => {
-            image: file,
-            copyright_holder: "zuul",
-            when: Time.current
-          }
-        }
-      }
-    )
+    Image.stub(:new, image) do
+      image.stub(:process_image, false) do
+        post(:create, params: params)
+      end
+    end
 
-    # Prove that an image was created, but that it is unattached, is in the
+    # Prove that an image was created, but is unattached, is in the
     # @bad_images array, and has not been kept in the @good_images array
     # for attachment later.
-    img = Image.find_by(copyright_holder: "zuul")
+
+    # img = Image.find_by(copyright_holder: "zuul")
+    img = Image.order(updated_at: :asc).last
```

Comment:
This was the only way I could figure out around the following problem.

Rails stubs `image` per lines 2870-2871, **except that it seems to ignore `copyright_holder: "zuul"`**.
In particular, if I put a breakpoint immediately after `app/controllers/observations_controller/form_helpers.rb` line 153
```ruby
          image = Image.new(args2.permit(permitted_image_args))
```
and examine `image`, I see
```ruby
(rdbg) l    # list command
   ...
   153|           image = Image.new(args2.permit(permitted_image_args))
=> 154|           debugger
   ...
(rdbg) image
#<Image:0x00007f90db78cfd0
 id: 1062212617,
 created_at: Sat, 14 Oct 2023 23:34:38.000000000 EDT -04:00,
 updated_at: Sat, 14 Oct 2023 23:34:38.000000000 EDT -04:00,
 content_type: nil,
 user_id: 241228755,
 when: Sat, 14 Oct 2023,
 notes: "stubbed in test",
 copyright_holder: "Rolf Singer",
 license_id: 982438718,
 num_views: 0,
 last_view: nil,
 width: nil,
 height: nil,
 vote_cache: nil,
 ok_for_export: true,
 original_name: "",
 transferred: false,
 gps_stripped: false,
 diagnostic: true>
```
I know this is the stubbed image because it includes the `notes` I created.
But why does it ignore the `copyright_holder` that I provided?

Reply:
So this is a weird one and it took a while for me to hunt it down, but the problem has to do with `init_new_image_var` in app/controllers/observations_controller/form_helpers.rb.  It is calling `Image.new` after it has been stubbed out and then it puts some default values into some of the fields.  The expectation is that these will get overridden by the params in a later call to `Image.new` at line 153 of the same file, but instead it just returns the now modified image.  In essence `Image.new` is getting called multiple times in the process for different reasons.  One potential way through it would be to change `init_new_image_var` so it takes a set of parameters rather than updating then after the call to `Image.new`.

Reply:
It looks like the above either wasn't understood or wasn't followed up on.  To be more concrete, I think we should try something like:
```
  def init_new_image_var(default_date)
    @new_image = Image.new(when: default_date, license: @user.license, copyright_holder: @user.legal_name)
  end
```
I haven't tested the above, but I think it's point towards something that should work. The idea is to get the stub of Image.new to return what we expect rather than changing what we return in the subsequent lines of code.